bson/src/main/org/bson/BsonDocument.java

```diff
@@ -735,10 +735,6 @@ public BsonValue put(final String key, final BsonValue value) {
         if (value == null) {
             throw new IllegalArgumentException(format("The value for key %s can not be null", key));
         }
-        if (key.contains("\0")) {
```

Comment:
> My understanding is that with this change it is now allowed to create a BsonDocument that violate the BSON specification.

Yes

> Have this already been the case regarding some other checks, of was it previously the case that if I have a BsonDocument, it is guaranteed to be writable to BSON?

It wasn't the case, since the checks are not include in BsonRegularExpression

> If there was an explicit decision to allow creating BsonDocuments that violate the BSON specification, then it's all fine, of course.

Essentially there was, since many drivers just rely on native dictionaries to represent BSON documents.

Including the check doesn't improve safety, since we have the check where it's really needed.  And there is a small performance cost to including it since we have to scan every key to check. It arguably improves things for users to include the check since they will find their mistake earlier, but it's still a runtime error any way you slice it.

For all these reasons I'm comfortable with this change.  I wouldn't have done it just to make it so we can run the new tests, but it's a nice benefit.

Reply:
Thanks for explaining 👍